src/managedwriter/encoder.ts

```diff
@@ -0,0 +1,133 @@
+const {Type} = protobuf;
+
+/**
+ * Internal class used by the JSONWriter to convert JSON data to protobuf messages.
+ * It can be configure to do some data conversion to match what BigQuery expects.
+ *
+ * @class
+ * @memberof managedwriter
+ */
+export class JSONEncoder {
+  private _type: protobuf.Type = Type.fromJSON('root', {
+    fields: {},
+  });
+
+  /**
+   * Creates a new JSONEncoder instance.
+   *
+   * @param {Object} params - The parameters for the JSONEncoder.
+   * @param {IDescriptorProto} params.protoDescriptor - The proto descriptor
+   *   for the JSON rows.
+   */
+  constructor(params: {protoDescriptor: IDescriptorProto}) {
+    const {protoDescriptor} = params;
+    this.setProtoDescriptor(protoDescriptor);
+  }
+
+  /**
+   * Update the proto descriptor for the Encoder.
+   *
+   * @param {IDescriptorProto} protoDescriptor - The proto descriptor.
+   */
+  setProtoDescriptor(protoDescriptor: IDescriptorProto): void {
+    const normalized = normalizeDescriptor(
+      new DescriptorProto(protoDescriptor)
+    );
+    this._type = Type.fromDescriptor(normalized);
+  }
+
+  /**
+   * Writes a JSONList that contains objects to be written to the BigQuery table by first converting
+   * the JSON data to protobuf messages, then using Writer's appendRows() to write the data at current end
+   * of stream. If there is a schema update, the current Writer is closed and reopened with the updated schema.
+   *
+   * @param {JSONList} rows - The list of JSON rows.
+   * @returns {Uint8Array[]} The encoded rows.
+   */
+  encodeRows(rows: any[]): Uint8Array[] {
+    const serializedRows = rows
+      .map(r => {
+        return this.convertRow(r);
+      })
+      .map(r => {
+        return this.encodeRow(r);
+      });
+    return serializedRows;
+  }
+
+  private isPlainObject(value: any): boolean {
+    return value && [undefined, Object].includes(value.constructor);
+  }
+
+  private encodeRow(row: any): Uint8Array {
+    const msg = this._type.create(row);
+    return this._type.encode(msg).finish();
+  }
+
+  private convertRow(source: any): Object {
+    const row = extend(true, {}, source);
+    for (const key in row) {
+      const value = row[key];
+      if (value === null) {
+        continue;
+      }
+      if (value instanceof Date) {
+        const pfield = this._type.fields[key];
+        if (!pfield) {
+          continue;
+        }
+        switch (pfield.type) {
+          case 'int32': // DATE
+            // The value is the number of days since the Unix epoch (1970-01-01)
+            row[key] = value.getTime() / (1000 * 60 * 60 * 24);
+            break;
+          case 'int64': // TIMESTAMP
+            // The value is given in microseconds since the Unix epoch (1970-01-01)
+            row[key] = value.getTime() * 1000;
+            break;
+          case 'string': // DATETIME
+            row[key] = value.toJSON().replace(/^(.*)T(.*)Z$/, '$1 $2');
```

Comment:
maybe `.toISOString()` will show the intention better

Reply:
`.toISOString()` has the T and Z on the string too and the timestamp literal on BQ require to remove that.